doc/ctlinnd.8
.\" $Revision$
.TH CTLINND 8
.SH NAME
ctlinnd \- control the InterNetNews daemon
.SH SYNOPSIS
.B ctlinnd
[
.B \-h
]
[
.B \-s
]
[
.BI \-t " timeout"
]
.I command
[
.I argument...
]
.SH DESCRIPTION
.I Ctlinnd
sends a message to the control channel of
.IR innd (8),
the InterNetNews server.
.PP
In the normal mode of behavior, the message is sent to the server, which
then performs the requested action and sends back a reply with a text
message and the exit code for
.IR ctlinnd .
If the server successfully performed the command,
.I ctlinnd
will exit with a status of zero and print the reply on standard output.
If the server could not perform the command (for example, it was told to
remove a newsgroup that does not exist), it will direct
.I ctlinnd
to exit with a status of one.
The ``shutdown,'' ``xabort,'' and ``xexec'' commands do not generate a reply;
.I ctlinnd
will always exit silently with a status of zero.
.SH OPTIONS
.TP
.B \-s
If the ``\fB\-s\fP'' flag is
used, then no message will be printed if the command was successful.
.TP
.B \-t
The ``\fB\-t\fP'' flag can be used to specify how long to wait for the reply
from the server.
The timeout value specifies the number of seconds to wait.
A value of zero waits forever, and a value less
than zero indicates that no reply is needed.
When waiting for a reply,
.I ctlinnd
will try every two minutes to see if the server is still running, so it
is unlikely that ``\fB\-t0\fP'' will hang.
The default is set as 
.I <CTLINND_TIMEOUT in config.data> 
(typically 
.\"=()<.IR @<typCTLINND_TIMEOUT>@ .)>()=
.IR 0 .)
.TP
.B \-h
To see a command summary, use the ``\fB\-h\fP'' flag.
If a command is included when
.I ctlinnd
is invoked with the ``\fB\-h\fP'' flag, then only the usage for that command
will be given.
.PP
If a large number of groups are going to be created or deleted at once,
it may be more efficient to ``pause'' or ``throttle'' the server
and edit the
.I active
file directly.
.PP
The complete list of commands follows.
Note that all commands have a fixed number of arguments.
If a parameter can be an empty string, then it is necessary to
specify it as two adjacent quotes, like "".
.TP
.BI addhist " <Message-ID> arr exp post paths"
Add an entry to the history database.
This directs the server to create a history line for
.IR Message-ID .
The angle brackets are optional.
.IR Arr ,
.IR exp ,
and
.I post
specify when the article arrived, what its expiration date is, and
when it was posted.
All three values are a number indicating the number of seconds since the
epoch.
If the article does not have an Expires header, then
.I exp
should be zero.
.I Paths
is the pathname within the news spool directory where the article is filed.
If the article is cross-posted, then the names should be separated by
whitespace and the
.I paths
argument should be inside double quotes.
If the server is throttled manually, this command causes it to briefly
open the history database.
And if the server is paused or throttled except that case, this command
is rejected.
.TP
.BI allow " reason"
Remote connections are allowed.
The
.I reason
must be the same text given with an earlier ``reject'' command, or an
empty string.
.TP
.BI begin " site"
Begin feeding
.IR site .
This will cause the server to rescan the
.IR newsfeeds (5)
file to find the specified site and set up a newsfeed for it.
If the site already exists, a ``drop'' is done first.
This command is forwarded; see below.
.TP
.BI cancel " <Message-ID>"
Remove the article with the specified Message-ID from the local system.
This does
.I not
generate a cancel message.
The angle brackets are optional.
If the server is throttled manually, this command causes it to briefly
open the history database.
And if the server is paused or throttled except that case, this command
is rejected.
.TP
.BI changegroup " group rest"
The newsgroup
.I group
is changed so that its fourth field in the
.I active
file becomes the value specified by the
.I rest
parameter.
This may be used to make an existing group moderated or unmoderated,
for example.
.TP
.B checkfile
Check the syntax of the
.I newsfeeds
file, and display a message if any errors are found.
The details of the errors are reported to
.IR syslog (3).
.TP
.BI drop " site"
Flush and drop
.I site
from the server's list of active feeds.
This command is forwarded; see below.
.TP
.BI feedinfo " site"
Print detailed information about the state of the
feed to
.I site
or more brief status of all feeds if
.I site
is an empty string.
.TP
.BI flush " site"
Flush the buffer for the specified site.
The actions taken depend on the type of feed the site receives; see
.IR newsfeeds (5).
This is useful when the site is fed by a file and batching is going to start.
If
.I site
is an empty string, then all sites are flushed and the
.I active
file and history databases are also written out.
This command is forwarded; see below.
.TP
.B flushlogs
Close the log and error log files and rename them to have a
.I \&.old
extension.
The history database and
.I active
file are also written out.
.TP
.BI go " reason"
Re-open the history database and start accepting articles after a ``pause''
or ``throttle'' command.
The
.I reason
must either be an empty string or match the text that was given
in the earlier ``pause'' or ``throttle'' command.
If a ``reject'' command was done, this will also do an ``allow'' command
if the
.I reason
matches the text that was given in the ``reject.''
If a ``reserve'' command was done, this will also clear the reservation if
the
.I reason
matches the text that was given in the ``reserve.''
Note that if only the history database has changed while the server is
paused or throttled, it is not necessary to send it a ``reload'' command
before sending it a ``go'' command.
If the server throttled itself because it accumulated too many I/O
errors, this command will reset the error count.
If the server was not started with the ``\fB\-ny\fP'' flag, then this command also
does a ``readers'' command with ``yes'' as the flag and
.I reason
as the text.
.TP
.BI hangup " channel"
Close the socket on the specified incoming channel.
This is useful when an incoming connection appears to be hung.
.TP
.BI help " [command]"
Print a command summary for all commands, or just
.I command
if specified.
.TP
.BI lowmark " file"
Reset the lowmarks in the active file based on the contents of the given
file. Each line in the file must be of the form:
.IP
.RS
.nf
    group low-value
.fi
.RE
.IP
for example
.IP
.RS
.nf
    comp.lang.c++    243
.fi
.RE
.IP
A useful perl script for this is (called ``findlow''):
.IP
.RS
.nf
    #!/usr/bin/perl
    require 5.001 ;
    require "/var/news/etc/innshellvars.pl" ;
    chdir $inn::spool or die "cd ($inn::spool): $!\n" ;
    open(A,"<$inn::active") or die "open ($inn::active): $!\n";
    while (<A>) {
		($g,$HI,$LO,$f) = split;
		$lo = $HI+1;
		$d = $g; $d =~ y/./\//;
		opendir(D, $d) or next;
		while ($f = readdir(D)) {
			next unless $f =~ /^\d+$/;
			$l = int($f);
			if ($l < $lo) { $lo = $l; }
		}
		closedir(D);
		print "$g $lo\n" if $lo != $LO;
    }
    close(A);
.fi
.RE
.IP
Note: the output only mentions groups who's lowmark doesn't match the active
file value. And a driver shell script is:
.IP
.RS
.nf
    T=/tmp/lowmark.$$
    if /usr/news/local/findlow > $T && [ -s $T ]; then
	ctlinnd -s lowmark $T
    fi
    rm $T
.fi
.RE
.TP
.BI logmode
Cause the server to log its current operating mode to
.IR syslog .
.TP
.BI mode
Print the server's operating mode as a multi-line summary of the parameters
and operating state.
.TP
.BI name " nnn"
Print the name of channel number
.I nnn
or of all channels if it is an empty string.
.TP
.BI newgroup " group rest creator"
Create the specified newsgroup.
The
.I rest
parameter should be the fourth field as described in
.IR active (5);
if it is not an equal sign, only the first letter is used.
The
.I creator
should be the name of the person creating the group.
If the newsgroup already exists, this is equivalent to the ``changegroup''
command.
This is the only command that has defaults.
The
.I creator
can be omitted and will default to the newsmaster (as specified at configure
time, ``usenet'' by default), and the
.I rest
parameter can be omitted and will default to ``y''.
This command can be done while the server is only throttled manually
or running; it will
update its internal state when a ``go'' command is sent.
This command updates the
.I active.times
(see
.IR active (5))
file.
This command is forwarded; see below.
.TP
.BI param " letter value"
Change the command-line parameters of the server.
The combination of defaults make it possible to use the text of the Control
header directly.
.I Letter
is the
.I innd
command-line option to set, and
.I value
is the new value.
For example, ``i 5'' directs the server to allow only five incoming
connections.
To enable or disable the action of the ``\fB\-n\fP'' flag, use the letter ``y''
or ``n'', respectively, for the
.IR value.
.TP
.BI pause " reason"
Pause the server so that no incoming articles are accepted.
No existing connections are closed, but the history database is closed.
This command should be used for short-term locks, such as when replacing
the history files.
If the server was not started with the ``\fB\-ny\fP'' flag, then this command also
does a ``readers'' command with ``no'' as the flag and
.I reason
as the text.
.TP
.BI perl " flag"
Enable or disable perl news filtering, if
.IR <\-\-with\-perl\ is\ specified\ at\ configure> .
If
.I flag
starts with the letter ``y'' then filtering is enabled.  If it starts with
``n'', then filtering is disabled.

.TP
.BI readers " flag text"
Allow or disallow newsreaders.
If
.I flag
starts with the letter ``n'' then newsreading is disallowed, by
causing the server to pass the
.I text
as the value of the
.IR nnrpd (8)
\&`\fB`\-r\fP'' flag.
If
.I flag
starts with the letter ``y'' and
.I text
is either an empty string, or the same string that was used when newsreading
was disallowed, then newsreading will be allowed.
.\".TP
.\".BI refile " path group"
.\"The article specified by
.\".I path
.\"is refiled as if it were posted to the newsgroup
.\".IR group .
.TP
.BI reject " reason"
Remote connections (those that would not be handed off to
.IR nnrpd )
are rejected, with
.I reason
given as the explanation.
.TP
.BI reload " what reason"
The server updates its in-memory copies of various configuration files.
.I What
identifies what should be reloaded.
If it is an empty string or the word ``all'' then everything is reloaded;
if it is the word ``history'' then the history database is closed and opened,
if it is the word ``incoming.conf'' then the
.IR incoming.conf (5)
file is reloaded; if it is the word ``active'' or ``newsfeeds'' then both
the
.I active
and
.I newsfeeds
files are reloaded; if it is the word ``overview.fmt'' then the
.IR overview.fmt (5)
file is reloaded; if it is the word ``overview.ctl'' then the
.IR overview.ctl (5)
file is reloaded.
If
.I <\-\-with\-perl is specified at configure>
and it is the word ``filter.perl'' then the
.IR filter_innd.pl
file is reloaded.  If a Perl procedure named ``filter_before_reload'' exists,
it will be called prior to rereading
.IR filter_innd.pl .
If a Perl procedure named ``filter_after_reload'' exists, it will be called
after
.IR filter_innd.pl .
has been reloaded.  Reloading the Perl filter does not enable filtering if
it is disabled; use
.IR perl y
to do this. The
.IR startup_innd.pl
file cannot be reloaded.

If
.I <\-\-with\-tcl is specified at configure>
and it is the word ``filter.tcl'' then the
.IR filter.tcl
file is reloaded.  If a TCL procedure named ``filter_before_reload'' exists,
it will be called prior to rereading
.IR filter.tcl.
If a TCL procedure named ``filter_after_reload'' exists, it will be called
after
.IR filter.tcl
has been reloaded.  Reloading the Tcl filter does not enable filtering if
it is disabled; use
.IR filter
to do this.
The
.I reason
is reported to
.IR syslog .
There is no way to reload the data
.IR inn.conf (5)
file.
The
.IR startup.tcl
file cannot be reloaded.
.TP
.BI renumber " group"
Scan the spool directory for the specified newsgroup and update the
low-water mark and hi-water mark in the
.I active
file.
If
.I group
is an empty string then all newsgroups are scanned.
.TP
.BI reserve " reason"
The next ``pause'' or ``throttle'' command must use
.I reason
as its reason.
This ``reservation'' is cleared by giving an empty string for the
.IR reason .
This command is used by programs like
.IR expire (8)
that want to avoid running into other instances of each other.
.TP
.BI rmgroup " group"
Remove the specified newsgroup.
This is done by editing the
.I active
file.
The spool directory is not touched, and any articles in the group will
be expired using the default expiration parameters.
Unlike the ``newgroup'' command, this command does not update the
.I active.times
file.
This command can be done while the server is only throttled manually or running.
This command is forwarded; see below.
.TP
.BI send " feed text..."
The specified
.I text
is sent as a control line to the exploder
.IR feed .
.TP
.BI shutdown " reason"
The server is shut down, with the specified reason recorded in the log
and sent to all open connections.
It is a good idea to send a ``throttle'' command first.
.TP
.BI signal " sig site"
Signal
.I sig
is sent to the specified
.IR site ,
which must be a channel or exploder feed.
.I Sig
can be a numeric signal number or the word ``hup,'' ``int,'' or ``term'';
case is not significant.
.TP
.BI tcl " flag"
Enable or disable Tcl news filtering, if
.IR <\-\-with\-tcl\ is\ specified\ at\ configure> .
If
.I flag
starts with the letter ``y'' then filtering is enabled.  If it starts with
``n'', then filtering is disabled.

.TP
.BI throttle " reason"
Input is throttled so that all existing connections are closed and new
connections are rejected.
The history database is closed.
This should be used for long-term locks, such as when
.I expire
is being run.
If the server was not started with the ``\-ny'' flag, then this command also
does a ``readers'' command with ``no'' as the flag and
.I reason
as the text.
.TP
.BI timer " off|interval"
Performance monitoring is turned off if 'off' or '0' is specified,
otherwise, statistics will be reported every
.I interval
seconds to syslog.  See 
.IR inn.conf (5)
for information on how to set the startup default.
.TP
.BI trace " item flag"
Tracing is turned on or off for the specified
.IR item .
.I Flag
should start with the letter ``y'' or ``n'' to turn tracing on or off.
If
.I item
starts is a number, then tracing is set for the specified
.I innd
channel, which must be for an incoming NNTP feed.
If it starts with the letter ``i'' then general
.I innd
tracing is turned on or off.
If it starts with the letter ``n'' then future
.IR nnrpd 's
will or will not have the ``\-t'' flag enabled, as appropriate.
.TP
.BI xabort " reason"
The server logs the specified
.I reason
and then invokes the
.IR abort (3)
routine.
.TP
.BI xexec " path"
The server gets ready to shut itself down, but instead of exiting it
execs
.I <PREFIX specified with \-\-prefix at configure>/inndstart
with all of its original arguments except for ``\fB\-r\fP''.
.I Path
can be any of ``innd'', ``inndstart'' or an empty string.
any other value is an error.
.PP
In addition to being acted upon within the server, certain commands can
be forwarded to the appropriate child process.
If the site receiving the command is an exploder (such as
.IR buffchan (8))
or it is a funnel that feeds into an exploder, then the
command can be forwarded.
In this case, the server will send a command line to the exploder that
consists of the
.I ctlinnd
command name.
If the site funnels into an exploder that has an asterisk (``*'') in its ``W''
flag (see
.IR newsfeed (5)),
then the site name will be appended to the command; otherwise no argument
is appended.
.SH BUGS
.I Ctlinnd
uses the
.IR inndcomm (3)
library, and is therefore limited to server replies no larger than 4k.
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> for InterNetNews.
.de R$
This is revision \\$3, dated \\$4.
..
.R$ $Id$
.SH "SEE ALSO"
active(5),
expire(8),
innd(8),
inndcomm(3),
inn.conf(5),
newsfeeds(5),
overview.fmt(5).


NEWS
Current

    Here are the changes for the latest version of INN.

    * New readers.conf file (replaces nnrp.access) which allows more
      flexible specification of access restrictions.

    * Unified overview has been replaced with an overview API, and there are
      now two separate overview implementations to choose from.  One
      (tradindexed) is very like traditional overview but uses an additional
      index file for faster responses to readers.  The other (buffindexed)
      uses circular buffers and can handle a higher incoming article rate
      while still being fast for readers.  The ovmethod key in inn.conf
      chooses the overview method to use.

    * All article storage and retrieval is now done via the storage API. 
      Traditional spool is now available as a storage type under the storage
      API.

    * The timecaf storage method has been added, similar to timehash but
      storing multiple articles in a single file.  See INSTALL for details
      on it.

    * INN now supports embedded Python filters as well as Perl and TCL
      filters.

Version 2.2

    * New storage.conf file (replaces storage.ctl).

    * New (optional) way of handling non-cancel control messages
      (controlchan) that serializes them and prevents server overload from
      control message storms.

    * Support for actsyncd to fetch active file with ftp; configured by
      default to get the one at ftp://ftp.isc.org/pub/usenet/CONFIG/active.Z
      if you run actsyncd.  Be sure to read the manual page for actsync to
      configure an actsync.ign file for your site, and test simpleftp if you
      do not configure with wget or ncftp.  Also see
      ftp://ftp.isc.org/pub/usenet/CONFIG/README.

    * Some options to configure are now moved to inn.conf (merge-to-groups
      and pgp-verify).

    * inndf, a portable version of df(1) is supplied.

    * New cnfsstat program to show stats of cnfs buffers.

    * news2mail program for gateway'ing news into email is provided.

    * mailpost program for gateway'ing email to news is provided.

    * pullnews program for doing a sucking feed is provided (not meant for
      large feeds).

    * The innshellvars.csh.in script is obsolete (and lives in the obsolete
      directory, for now).



frontends/ctlinnd.c
/*  $Revision$
**
**  Send control messages to the InterNetNews daemon.
*/
#include <stdio.h>
#include <sys/types.h>
#include "configdata.h"
#include "clibrary.h"
#include <ctype.h>
#include <errno.h>
#include <sys/stat.h>
#include "libinn.h"
#include "inndcomm.h"
#include "paths.h"
#include "macros.h"


/*
**  Datatype for an entry in the command table.
*/
typedef struct _COMMAND {
    STRING	Command;
    STRING	Text;
    int		argc;
    char	Letter;
    BOOL	Glue;
} COMMAND;


STATIC COMMAND	Commands[] = {
    {	"addhist",	"id arr exp post path...\tAdd history line",
	5,	SC_ADDHIST,	TRUE	},
    {	"allow",	"reason...\t\t\tAllow remote connections",
	1,	SC_ALLOW,	TRUE	},
    {	"begin",	"site\t\t\tStart newly-added site",
	1,	SC_BEGIN,	FALSE	},
    {	"cancel",	"id\t\t\tCancel message locally",
	1,	SC_CANCEL,	FALSE	},
    {	"changegroup",	"group rest\tChange mode of group",
	2,	SC_CHANGEGROUP,	FALSE	},
    {	"checkfile",	"\t\t\tCheck syntax of newsfeeds file",
	0,	SC_CHECKFILE,	FALSE	},
    {	"drop",		"site\t\t\tStop feeding site",
	1,	SC_DROP,	FALSE		},
    {	"feedinfo",		"site\t\t\tPrint state of feed to site*",
	1,	SC_FEEDINFO,	FALSE		},
#if defined(DO_TCL)
    {	"tcl",			"flag\t\t\tEnable or disable Tcl filtering",
	1,	SC_FILTER,	FALSE		},
#endif /* defined(DO_TCL) */
    {	"flush",	"site\t\t\tFlush feed for site*",
	1,	SC_FLUSH,	FALSE	},
    {	"flushlogs",	"\t\t\tFlush log files",
	0,	SC_FLUSHLOGS,	FALSE	},
    {	"go",		"reason...\t\t\tRestart after pause or throttle",
	1,	SC_GO,		TRUE	},
    {	"hangup",	"channel\t\tHangup specified incoming channel",
	1,	SC_HANGUP,	FALSE	},
    {	"logmode",		"\t\t\t\tSend server mode to syslog",
	0,	SC_LOGMODE,	FALSE		},
    {	"mode",		"\t\t\t\tPrint operating mode",
	0,	SC_MODE,	FALSE		},
    {	"name",		"nnn\t\t\tPrint name of specified channel*",
	1,	SC_NAME,	FALSE		},
    {	"newgroup",	"group rest creator\tCreate new group",
	3,	SC_NEWGROUP,	FALSE	},
    {	"param",	"letter value\t\tChange command-line parameters",
	2,	SC_PARAM,	FALSE	},
    {	"pause",	"reason...\t\tShort-term pause in accepting articles",
	1,	SC_PAUSE,	TRUE	},
#if defined(DO_PERL)
    {	"perl",			"flag\t\t\tEnable or disable Perl filtering",
	1,	SC_PERL,	FALSE	},
#endif /* defined(DO_PERL) */
    {	"readers",	"flag text...\t\tEnable or disable newsreading",
	2,	SC_READERS,	TRUE	},
    {	"reject",	"reason...\t\t\tReject remote connections",
	1,	SC_REJECT,	TRUE	},
    {	"reload",	"what reason...\t\tRe-read config files*",
	2,	SC_RELOAD,	TRUE	},
    {	"renumber",	"group\t\tRenumber the active file*",
	1,	SC_RENUMBER,	FALSE	},
    {	"reserve",	"reason...\t\tReserve the next pause or throttle",
	1,	SC_RESERVE,	TRUE	},
    {	"rmgroup",	"group\t\t\tRemove named group",
	1,	SC_RMGROUP,	FALSE	},
    {	"send",		"feed text...\t\tSend text to exploder feed",
	2,	SC_SEND,	TRUE	},
    {	"shutdown",	"reason...\t\tShut down server",
	1,	SC_SHUTDOWN,	TRUE	},
    {	"kill",	"signal site\t\tSend signal to site's process",
	2,	SC_SIGNAL,	FALSE	},
    {	"throttle",	"reason...\t\tStop accepting articles",
	1,	SC_THROTTLE,	TRUE	},
    {   "timer",        "interval|off\t\tTurn performance monitoring on or off",
	1,      SC_TIMER,      FALSE    },
    {	"trace",	"innd|#|nnrpd flag\tTurn tracing on or off",
	2,	SC_TRACE,	FALSE	},
    {	"xabort",	"text...\t\tAbort the server",
	1,	SC_XABORT,	TRUE	},
    {	"xexec",	"path\t\t\tExec new server",
	1,	SC_XEXEC,	FALSE	}
};



/*
**  Print a help summary.
*/
STATIC NORETURN
Help(p)
    char		*p;
{
    register COMMAND	*cp;

    if (p == NULL) {
	(void)printf("Command summary:\n");
	for (cp = Commands; cp < ENDOF(Commands); cp++)
	    (void)printf("  %s %s\n", cp->Command, cp->Text);
	(void)printf("*   Empty string means all sites/groups/etc.\n");
	(void)printf("... All trailing words are glued together.\n");
	exit(0);
    }
    for (cp = Commands; cp < ENDOF(Commands); cp++)
	if (EQ(p, cp->Command)) {
	    (void)printf("Command usage:\n");
	    (void)printf("  %s %s\n", cp->Command, cp->Text);
	    exit(0);
	}
    (void)printf("No such command.\n");
    exit(0);
}


/*
**  Print a command-usage message and exit.
*/
STATIC NORETURN
WrongArgs(cp)
    COMMAND	*cp;
{
    (void)printf("Wrong number of arguments -- usage:\n");
    (void)printf("  %s %s\n", cp->Command, cp->Text);
    exit(1);
}


/*
**  Print an error message and exit.
*/
STATIC NORETURN
Failed(p)
    char	*p;
{
    if (ICCfailure)
	(void)fprintf(stderr, "Can't %s (%s failure) %s.\n",
		p, ICCfailure, strerror(errno));
    else
	(void)fprintf(stderr, "Can't %s, %s.\n", p, strerror(errno));
    (void)ICCclose();
    exit(1);
}


/*
**  Print an error reporting incorrect usage.
*/
STATIC NORETURN
Usage(what)
    char	*what;
{
    (void)fprintf(stderr, "Usage error (%s) -- try -h for help.\n", what);
    exit(1);
}


int main(int ac, char *av[])
{
    static char		Y[] = "y";
    static char		EMPTY[] = "";
    COMMAND	        *cp;
    char	        *p;
    int	                i;
    BOOL		Silent;
    BOOL		NeedHelp;
    char		*reply;
    char		*new;
    int			length;
    char		*nv[4];
    struct stat		Sb;
    char		buff[SMBUF];

    /* Set defaults. */
    Silent = FALSE;
    NeedHelp = FALSE;
    ICCsettimeout(CTLINND_TIMEOUT);

    /* Parse JCL. */
    while ((i = getopt(ac, av, "hst:")) != EOF)
	switch (i) {
	default:
	    Usage("bad flags");
	    /* NOTREACHED */
	case 'h':		/* Get help			*/
	    NeedHelp = TRUE;
	    break;
	case 's':		/* Silent -- no output		*/
	    Silent = TRUE;
	    break;
	case 't':		/* Time to wait for reply	*/
	    ICCsettimeout(atoi(optarg));
	    break;
	}
    ac -= optind;
    av += optind;
    if (NeedHelp)
	Help(av[0]);
    if (ac == 0)
	Usage("missing command");

    /* Look up the command word and move to the arguments. */
    if (EQ(av[0], "help"))
	Help(av[1]);
    for (cp = Commands; cp < ENDOF(Commands); cp++)
	if (EQ(av[0], cp->Command))
	    break;
    if (cp == ENDOF(Commands))
	Usage("unknown command");
    ac--;
    av++;

    /* Check argument count. */
    if (cp->Letter == SC_NEWGROUP) {
	/* Newgroup command has defaults. */
	switch (ac) {
	default:
	    WrongArgs(cp);
	    /* NOTREACHED */
	case 1:
	    nv[0] = av[0];
	    nv[1] = Y;
	    nv[2] = EMPTY;
	    nv[3] = NULL;
	    av = nv;
	    break;
	case 2:
	    nv[0] = av[0];
	    nv[1] = av[1];
	    nv[2] = EMPTY;
	    nv[3] = NULL;
	    av = nv;
	    break;
	case 3:
	    break;
	}
	switch (av[1][0]) {
	default:
	    Usage("Bad group mode");
	    /* NOTREACHED */
	case NF_FLAG_ALIAS:
	case NF_FLAG_EXCLUDED:
	case NF_FLAG_MODERATED:
	case NF_FLAG_OK:
	case NF_FLAG_NOLOCAL:
	case NF_FLAG_IGNORE:
	    break;
	}
	ac = 3;
    }
    else if (ac > cp->argc && cp->Glue) {
	/* Glue any extra words together. */
	for (length = 0, i = cp->argc - 1; (p = av[i++]) != NULL; )
	    length += strlen(p) + 1;
	for (new = p = NEW(char, length), i = cp->argc - 1; av[i]; i++) {
	    if (i >= cp->argc)
		*p++ = ' ';
	    p += strlen(strcpy(p, av[i]));
	}
	av[cp->argc - 1] = new;
	av[cp->argc] = NULL;
    }
    else if (ac != cp->argc)
	/* All other commands must have the right number of arguments. */
	WrongArgs(cp);

    /* Make sure there are no separators in the parameters. */
    for (i = 0; (p = av[i++]) != NULL; )
	if (strchr(p, SC_SEP) != NULL) {
	    (void)fprintf(stderr, "Illegal character '\\%03o' in \"%s\"\n",
		    SC_SEP, p);
	    exit(1);
	}

    /* Do the real work. */
    if (ICCopen() < 0)
	Failed("setup communication");
    i = ICCcommand(cp->Letter, av, &reply);
    if (i < 0) {
	i = errno;
	if (stat(_PATH_SERVERPID, &Sb) < 0)
	    (void)fprintf(stderr, "No innd.pid file; did server die?\n");
	(void)sprintf(buff, "send \"%s\" command", cp->Command);
	errno = i;
	Failed(buff);
    }

    if (reply) {
	/* Skip "<exitcode><space>" part of reply. */
	for (p = reply; *p && CTYPE(isdigit, *p); p++)
	    continue;
	while (*p && ISWHITE(*p))
	    p++;
	if (i != 0)
	    (void)fprintf(stderr, "%s\n", p);
	else if (!Silent)
	    (void)printf("%s\n", p);
    }

    if (ICCclose() < 0)
	Failed("end communication");

    exit(i);
    /* NOTREACHED */
}
